### PR TITLE
win: implements uv__udp_try_send

### PR DESCRIPTION
**Pull request by gemabarni**
This PR implements the missing `uv__udp_try_send` method for Windows.

### CODE REVIEW

File: src/win/udp.c [hunk 1]
```diff
@@ -924,5 +924,44 @@ int uv__udp_try_send(uv_udp_t* handle,
...
+
+  DWORD bytes;
+  const struct sockaddr* bind_addr;
+  int err;
+
+  assert(nbufs > 0);
+
+  /* already sending a message */
+  if (handle->send_queue_count != 0)
+  return UV_EAGAIN;
```
Comment (cjihrig) [inline src/win/udp.c]: Indentation is off.

File: src/win/udp.c [hunk 2]
```diff
@@ -924,5 +924,44 @@ int uv__udp_try_send(uv_udp_t* handle,
...
+
+  if (!(handle->flags & UV_HANDLE_BOUND)) {
+    if (addrlen == sizeof(uv_addr_ip4_any_)) {
+      bind_addr = (const struct sockaddr*) &uv_addr_ip4_any_;
+    }
+    else if (addrlen == sizeof(uv_addr_ip6_any_)) {
```
Comment (cjihrig) [inline src/win/udp.c]: This should go on the same line as the previous curly brace. But, house style is to leave off the curly braces if all of the branches are a single statement.

File: src/win/udp.c [hunk 3]
```diff
@@ -924,5 +924,44 @@ int uv__udp_try_send(uv_udp_t* handle,
...
+    else {
+      abort();
+    }
+    err = uv_udp_maybe_bind(handle, bind_addr, addrlen, 0);
+    if (err)
+      return uv_translate_sys_error(err);
+  }
+
+  err = WSASendTo(handle->socket,
+    (WSABUF*)bufs,
```
Comment (cjihrig) [inline src/win/udp.c]: Can you line these arguments up with `handle->socket`.

**Comment by gemabarni** [comment]
Thanks @cjihrig for the feedback, fixed them.

File: src/win/udp.c [hunk 4]
```diff
@@ -924,5 +924,41 @@ int uv__udp_try_send(uv_udp_t* handle,
                      unsigned int nbufs,
                      const struct sockaddr* addr,
                      unsigned int addrlen) {
-  return UV_ENOSYS;
+
```
Comment (cjihrig) [inline src/win/udp.c]: This blank line can be removed.

**Comment by cjihrig** [comment]
Any Windows people care to give this a review?

File: src/win/udp.c [hunk 5]
```diff
@@ -924,5 +924,41 @@ int uv__udp_try_send(uv_udp_t* handle,
...
+  DWORD bytes;
+  const struct sockaddr* bind_addr;
+  int err;
+
+  assert(nbufs > 0);
+
+  /* already sending a message */
```
Comment (bnoordhuis) [inline src/win/udp.c]: Can you capitalize and punctuate the comment?

**Comment by txdv** [comment]
Maybe you should enable the [udp-try-send](https://github.com/libuv/libuv/blob/v1.x/test/test-udp-try-send.c#L29-L37) test in this PR?

**Comment by saghul** [comment]
@txdv He already did.

File: src/win/udp.c [hunk 6]
```diff
@@ -924,5 +924,40 @@ int uv__udp_try_send(uv_udp_t* handle,
...
+  if (handle->send_queue_count != 0)
+    return UV_EAGAIN;
+
+  if (!(handle->flags & UV_HANDLE_BOUND)) {
+    if (addrlen == sizeof(uv_addr_ip4_any_))
+      bind_addr = (const struct sockaddr*) &uv_addr_ip4_any_;
+    else if (addrlen == sizeof(uv_addr_ip6_any_))
+      bind_addr = (const struct sockaddr*) &uv_addr_ip6_any_;
+    else
+      abort();
```
Comment (saghul) [inline src/win/udp.c]: I'd return UV_EINVAL here.  What do we do in similar places in the code?

**Comment by saghul** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/210/

File: src/win/udp.c (hunk 6 above)
Comment (gemabarni) [inline src/win/udp.c]: @saghul uv__udp_send does the same, but I can rewrite it to return UV_EINVAL if you would like. Probably it makes more sense.

Comment (saghul) [inline src/win/udp.c]: Yeah, let's do that.

**Comment by santigimeno** [comment]
@gemabarni ping! :)

**Comment by cjihrig** [comment]
It looks like @supercairos has picked this up. Moving to #1385.

**Pull request closed without merging**